### ISSUE TITLE
Rasterizer SSE2 optimization

### ISSUE THREAD

**Issue by jry2** [memononen/nanosvg#43]
I tried SSE2 version of `nsvg__scanlineSolid()` with `NSVG_PAINT_COLOR` code path converted. Benchmark on my i5 661 @ 3.5GHz, Windows 7 x64, Visual Studio 2015 RC, x86 release target. Rendering Ghostscript_Tiger.svg, measuring `nsvgRasterize()` time.

Upstream NanoSVG
900x900px: 68ms
9000x9000px: 4256ms

SSE2 NanoSVG
900x900px: 60ms
9000x9000px: 3125ms

Some improvement, but nothing stellar. I didn't use SSE before so maybe someone more experienced could do better. Anyone interested in my quick&dirty patch? Output PNG is binary same for both upstream and sse2 versions. 

Streaming SIMD Extensions (/arch:SSE) option was enabled for whole application. There is another boost with Streaming SIMD Extensions 2 (/arch:SSE2) enabled, but there are still CPUs not supporting SSE2.

**Comment by memononen** [comment]
Nice! Have you checked on higher level how much time is spent in flattenPath, qsort, and rasterize sorted edges? I expect the rasterization to dominate, but just curious. Also, what is the proportion of nsvg__scanlineSolid of nsvg__rasterizeSortedEdges?

**Comment by jry2** [comment]
Yes, see attached screenshots from release (upstream) build. Rendering 9000x9000px.
`nsvg__unpremultiplyAlpha()` is another SSE2 candidate.

![tiger_release_profiler1](https://cloud.githubusercontent.com/assets/10026109/8148549/828041b6-12a2-11e5-8a2f-7f098470cf6e.png)
![tiger_release_profiler2](https://cloud.githubusercontent.com/assets/10026109/8148550/8283f6da-12a2-11e5-865a-eadc82360f27.png)
![tiger_release_profiler3](https://cloud.githubusercontent.com/assets/10026109/8148558/ce21b122-12a2-11e5-8001-da6e2c123d81.png)

**Comment by jry2** [comment]
Same options but rendering to 900x900 target.

![tiger_900_1](https://cloud.githubusercontent.com/assets/10026109/8148591/8fabb562-12a4-11e5-94c6-b693c53c41c1.png)

**Comment by bengarney** [comment]
You should do it in NEON! What does your patch look like?

**Comment by jry2** [comment]
I'm working on x86/x64 project for Windows so ARM-NEON would not help. I will publish my patch.

**Comment by jry2** [comment]
Commit: https://github.com/jry2/nanosvg/commit/20db7eb52c728d3898dc1fa20089a8f28c2d4e60

**Comment by jry2** [comment]
Another benchmark (Ghostscript_Tiger.svg rendered 9000x9000px), tested x86 vs x64 performance.

Upstream version
x86: 4120ms, x64: 2960ms

SSE2 version
x86: 3100ms, x64: 2270ms

**Comment by james2432** [comment]
this seems like a nice optimization ever consider creating a pull request to get this merged?

**Comment by DsoTsin** [comment]
you should do it with intel ispc

**Comment by cbum13** [comment]
Benchmark on my PC
- Intel Pentium G630 @ 2.7GHz
- Windows 8.1 x64
- clang version 7.1.0 (tags/RELEASE_710/final)
- x86 release target.

Rendering Ghostscript_Tiger.svg, measuring nsvgRasterize() time.

Upstream NanoSVG
900x900    - ~65ms
9000x9000  - ~4740ms

SSE2 Optimized NanoSVG
900x900    - ~32ms
9000x9000  - ~1130ms
